### PR TITLE
Add developer documentation

### PR DESCRIPTION
**Pull request by danielmellado**
This PR adds devel docs and a quickstart for Egress-Router-CNI.

### CODE REVIEW

**Comment by juanluisvaladas** [comment]
It looks good to me, but we know the oxen we plow with and neither of us is precisely Shakespeare, let's wait for Dan to review it.

File: .gitignore [hunk 1]
```diff
@@ -1 +1,28 @@
 bin/
+# Compiled Object files, Static and Dynamic libs (Shared Objects)
+*.o
+*.a
+*.so
```
Comment (danwinship) [inline .gitignore:5]: where did all of this stuff come from? was it autogenerated? Most (all?) of this is irrelevant

File: docs/content/post/developer-documentation.md [hunk 2]
```diff
@@ -0,0 +1,256 @@
...
+Chain POSTROUTING (policy ACCEPT)
+target     prot opt source               destination
+SNAT       all  --  anywhere             anywhere             to:192.168.10.99
+
+```
+
+## Using quay.io/openshift image
+
+In order to deploy this into a cluster, you'll have to either compile the
+inclusded `Dockerfile` or rely into the preexising images that are shipped into
```
Comment (danwinship) [inline docs/content/post/developer-documentation.md:112]: "included"

File: docs/content/post/developer-documentation.md [hunk 3]
```diff
@@ -0,0 +1,256 @@
...
+openshift's quay
+(https://quay.io/repository/openshift/origin-egress-router-cni).
+
+If deploying an openshift cluster, we're already deploying this image alongside
+CNO's multus' daemonsets, so you've got nothing to do. Only if you want to
+change the default image to yours, you'll have to override the
+`EGRESS_ROUTER_CNI_IMAGE` value with the reference to your image.
```
Comment (danwinship) [inline docs/content/post/developer-documentation.md:119]: Override where? If you mean CNO, you can't easily do that since CVO will undo the change to the CNO Deployment

File: docs/content/post/developer-documentation.md [hunk 4]
```diff
@@ -0,0 +1,256 @@
...
+
+Egress Router CNI aims to match openshift-sdn functionality, please refer to
+[this nice openshift blog article](https://www.openshift.com/blog/accessing-external-services-using-egress-router)
+for further details on it. In order to make things easier for the legacy router
+users, so far the `redirect` mode is supported in egress-router-cni.
+
+In openshift-sdn, the egress router was implemented by adding an annotation to
+allow a pod to request a macvlan interface. In order to avoid repeating this
+behavior in ovn-kubernetes, we'd be requesting such interface using multus to
+ensure feature-parity with openshift-sdn.
```
Comment (danwinship) [inline docs/content/post/developer-documentation.md:135]: "In openshift-sdn, the egress router was implemented by adding an annotation to
allow a pod to request a macvlan interface. In contrast, the new egress-router
implementation uses CNI to create the macvlan interface and so can be used
with any network plugin."

(ie, it's not an ovn-kubernetes feature. It's a network-plugin-agnostic feature.)

File: docs/content/post/developer-documentation.md [hunk 5]
```diff
@@ -0,0 +1,256 @@
...
+
+Please take a look at the [enhancement
+proposal](https://github.com/openshift/enhancements/blob/master/enhancements/network/egress-router.md)
+if you want to get some more detailed info on this.
+
+Assumming that you have the cni plugin installed, you'd be able to create a
```
Comment (danwinship) [inline docs/content/post/developer-documentation.md:141]:
```suggestion
Assumming that you have the CNI plugin installed, you'd be able to create a
```

File: docs/content/post/developer-documentation.md [hunk 6]
```diff
@@ -0,0 +1,256 @@
...
+    }'
+```
+
+This would create the additional network, which would be later used in the pod
+with the macvlan interface.
+
+Let's go create the pod!
+
+> NOTE:
+> A pod image with iptables is required.
```
Comment (danwinship) [inline docs/content/post/developer-documentation.md:186]: This is not true; egress-router-cni uses the system's iptables binaries, not anything from the pod image.

File: docs/content/post/developer-documentation.md [hunk 7]
```diff
@@ -0,0 +1,256 @@
...
+2: eth0@if6: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP group default
+    link/ether 1a:9b:80:a1:1e:99 brd ff:ff:ff:ff:ff:ff link-netnsid 0
+    inet 192.168.10.99/24 brd 192.168.10.255 scope global eth0
+       valid_lft forever preferred_lft forever
+    inet6 fe80::189b:80ff:fea1:1e99/64 scope link
+       valid_lft forever preferred_lft forever
+```
+
+```bash
+$ sudo ip netns exec testing iptables -t nat -L
```
Comment (danwinship) [inline docs/content/post/developer-documentation.md:92]: I recommend _always_ using `iptables-save`, not `iptables -L`

**Comment by danielmellado** [comment]
Pull-request updated, HEAD is now e4970eab16fbccb9101dd5536bfe5b6411efe159

**Comment by danielmellado** [comment]
Pull-request updated, HEAD is now d685aa9ae776d41b7d642a02a0602a52cacba3aa

**Comment by danielmellado** [comment]
Pull-request updated, HEAD is now 8acf7488a4ec6a7e01709240f3c59dea90e9acef

**Comment by danielmellado** [comment]
Pull-request updated, HEAD is now 5f6271341c454d02c63abe40f3c948c2b5b58247

File: docs/content/post/developer-documentation.md [hunk 8]
```diff
@@ -0,0 +1,310 @@
...
+
+Chain POSTROUTING (policy ACCEPT)
+target     prot opt source               destination
+SNAT       all  --  anywhere             anywhere             to:192.168.10.99
+
+```
+
+## Using quay.io/openshift image
+
+In order to deploy this into a cluster, you'll have to either compile the
```
Comment (astoycos) [inline docs/content/post/developer-documentation.md:111]: nit: "build the image locally with the provided `Dockerfile` or .... " Sounds a bit neater I think

File: docs/content/post/developer-documentation.md [hunk 9]
```diff
@@ -0,0 +1,310 @@
...
++++
+
+This article covers how to install and test the Egress Router CNI, both using
+cnitool or deploying openshift and using its official image.
+
+## Using cnitool
+
+[cnitool](https://github.com/containernetworking/cni/tree/master/cnitool) is
+a simple program that executes a CNI configuration. It will add or
+remove an interface in an already-created network namespace. You can use it in
```
Comment (astoycos) [inline docs/content/post/developer-documentation.md:20]: nit: Should we be using first, second or third person in this documentation? I'm not sure on the protocol here @jboxman, but it my mind wouldn't we be avoiding personal pronouns?

File: docs/content/post/developer-documentation.md [hunk 10]
```diff
@@ -0,0 +1,310 @@
...
+included `Dockerfile` or rely into the preexising images that are shipped into
+openshift's quay
+(https://quay.io/repository/openshift/origin-egress-router-cni).
+
+If deploying an openshift cluster, we're already deploying this image alongside
+CNO's multus' daemonsets, so you've got nothing to do. Only if you want to
+change the default image to yours, you'll have to override the
+`EGRESS_ROUTER_CNI_IMAGE` value with the reference to your image within
+`env.sh` file for CNO's `hack/run-locally`.
+
```
Comment (astoycos) [inline docs/content/post/developer-documentation.md:121]: nit: the wording here is a bit awk to me, Maybe something like 
"By default the egress router image is deployed within the CNO's multus daemonsets for all Openshift clusters.  However, If a custom image is required the `EGRESS_ROUTER_CNI_IMAGE` value can be specified in the 
`env.sh` file for CNO's `hack/run-locally`.

File: docs/content/post/developer-documentation.md [hunk 11]
```diff
@@ -0,0 +1,310 @@
...
+```
+
+## Using quay.io/openshift image
+
+In order to deploy this into a cluster, you'll have to either compile the
+included `Dockerfile` or rely into the preexising images that are shipped into
+openshift's quay
+(https://quay.io/repository/openshift/origin-egress-router-cni).
+
+If deploying an openshift cluster, we're already deploying this image alongside
```
Comment (astoycos) [inline docs/content/post/developer-documentation.md:116]: nit: shouldn't openshift -> Openshift?

File: docs/content/post/developer-documentation.md [hunk 12]
```diff
@@ -0,0 +1,310 @@
...
+`EGRESS_ROUTER_CNI_IMAGE` value with the reference to your image within
+`env.sh` file for CNO's `hack/run-locally`.
+
+If you don't want to use CNO at all, you can still install the egress router
+cni by manually copying it to the CNI binary path, which should be
+`/var/lib/cni/bin/`
+
+### How to use the egress-router-cni
+
+Egress Router CNI aims to match openshift-sdn functionality, please refer to
```
Comment (astoycos) [inline docs/content/post/developer-documentation.md:128]: nit: "The Egress Router CNI ..."

Comment (astoycos) [inline docs/content/post/developer-documentation.md:128]: nit "openshift-sdn" -> "openshift-sdn's"

File: docs/content/post/developer-documentation.md [hunk 13]
```diff
@@ -0,0 +1,310 @@
...
+
+(i.e. it's not an ovn-kubernetes. It's a network-plugin-agnostic feature.)
+
+Please take a look at the [enhancement
+proposal](https://github.com/openshift/enhancements/blob/master/enhancements/network/egress-router.md)
+if you want to get some more detailed info on this.
+
+Assumming that you have the CNI plugin installed, you'd be able to create a
+NetworkAttachmentDefinition (NAD) with the CNI configuration alongside, just
+such as in the below example.
```
Comment (astoycos) [inline docs/content/post/developer-documentation.md:146]: nit: remove "just such"

File: docs/content/post/developer-documentation.md [hunk 14]
```diff
@@ -0,0 +1,310 @@
...
+  containers:
+    - name: openshift-egress-router-pod
+      command: ["/bin/bash", "-c", "sleep 999999999"]
+      image: centos/tools
+      securityContext:
+        privileged: true
+
+```
+
+If we now check out the annotations from the just-created pod, we'd be able to
```
Comment (astoycos) [inline docs/content/post/developer-documentation.md:212]: nit: just-created -> new

File: docs/content/post/examples.md [hunk 15]
```diff
@@ -0,0 +1,120 @@
...
+      "gateway": "192.168.123.1"
+      }
+    }'
+```
+
+### NetworkAttachmentDefinition with CNI config file
+If NetworkAttachmentDefinition has no spec, multus find a file in
+defaultConfDir (‘/etc/cni/multus/net.d’, with same name in the ‘name’ field of
+CNI config.
+```yaml
```
Comment (astoycos) [inline docs/content/post/examples.md:43]: The wording here is a bit confusing, do you mean "If the NetworkAttatchmentDefinition has no spec, multus will look for a file in the defaultConfDir (‘/etc/cni/multus/net.d’), with same name that's specified in the CNI config."

**Review by astoycos — COMMENTED** [review]
Overall /lgtm just barring the few nits and questions!

File: docs/content/post/developer-documentation.md (hunk 9 above)
Comment (jboxman) [reply docs/content/post/developer-documentation.md:20]: @astoycos, it depends. I think outside of OpenShift docs, the style is undefined. For OCP docs, we do use _you_.
